onnxscript/rewriter/ort_fusions/fuse_packed_qkv_gqa_test.py

```diff
@@ -0,0 +1,147 @@
+from onnxscript import opset18 as op
+from onnxscript.rewriter.ort_fusions._test_utils import assert_allclose
+from onnxscript.rewriter.ort_fusions.fuse_packed_qkv_gqa import fuse_qkv_gqa
+
+msft_op = onnxscript.values.Opset("com.microsoft", 1)
+
+# Test case for fusion of separate query, key and value inputs
+# into a single packed QKV input for the GroupQueryAttention operator.
+
+
+class PackedQKVforGQAFusionTest(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Config parameters
+        self.batchsize = 1
+        self.seqlen = 8
+        self.kv_seqlen = self.seqlen
+        self.past_seqlen = 16
+        self.head_size = 16
+        self.q_num_heads = 20
+        self.kv_num_heads = 10
+
+        # Computed config parameters
+        self.q_hidden_size = self.head_size * self.q_num_heads
+        self.kv_hidden_size = self.head_size * self.kv_num_heads
+        self.hidden_size = self.q_hidden_size + self.kv_hidden_size + self.kv_hidden_size
+
+        # Abbreviations
+        B = self.batchsize
+        S = self.seqlen
+        P = self.past_seqlen
+        D = self.hidden_size
+        Dh = self.head_size
+        Hkv = self.kv_num_heads
+        total_seqlen = S + P
+        max_seqlen = total_seqlen
+
+        self.input_types = (
+            FLOAT["B", "S", D],  # packed_qkv
+            FLOAT["B", Hkv, "P", Dh],  # past_key
+            FLOAT["B", Hkv, "P", Dh],  # past_value
+            FLOAT["max_seqlen", Dh // 2],  # cos
+            FLOAT["max_seqlen", Dh // 2],  # sin
+        )
+        self.output_types = (
+            FLOAT["B", "S", D],  # attention
+            FLOAT["B", Hkv, "T", Dh],  # present_key
+            FLOAT["B", Hkv, "T", Dh],  # present_value
+        )
+
+        self.inputs = {
+            "packed_qkv": np.random.rand(B, S, D).astype(np.float32),
+            "past_key": np.random.rand(B, Hkv, P, Dh).astype(np.float32),
+            "past_value": np.random.rand(B, Hkv, P, Dh).astype(np.float32),
+            "cos": np.random.rand(max_seqlen, Dh // 2).astype(np.float32),
+            "sin": np.random.rand(max_seqlen, Dh // 2).astype(np.float32),
+        }
+
+    def source_model_script(self):
+        Hq = self.q_num_heads
+        Hkv = self.kv_num_heads
+
+        @script()
+        def gqa(packed_qkv, past_key, past_value, cos, sin):
+            # Generate seqlens_k and total_seqlen inputs for GQA:
+            # In this test case, all batch elements have same sequence length.
+            S = op.Shape(packed_qkv, start=1, end=2)
+            past_seq_length = op.Shape(past_key, start=2, end=3)
+            total_seq_length = op.Add(past_seq_length, S)
+            total_seqlen_int32 = op.Cast(total_seq_length, to=6)
+            total_seqlen_int32_minus_1 = op.Sub(total_seqlen_int32, 1)
+            batchsize = op.Shape(packed_qkv, start=0, end=1)
+            seqlens_k = op.Tile(total_seqlen_int32_minus_1, batchsize)
```

Comment:
I am curious why we need the above code fragment (for a minimal test case). Would it be possible to simply make seqlens_k and total_seqlen_int32 just be inputs of this model (just like cos, sin, etc.)? It's a nit, but seems better

Reply:
Done